stimpacks/deploy/config.go

```diff
@@ -0,0 +1,352 @@
+// Deployment describes details about the deployment assets (directories, files, etc)
+type Deployment struct {
+	Directory         string    `yaml:"directory"`
+	Script            string    `yaml:"script"`
+	Container         Container `yaml:"container"`
+	fullDirectoryPath string
+}
+
+// Container describes the container used for Docker deployments
+type Container struct {
+	Repo string `yaml:"repo"`
+	Tag  string `yaml:"tag"`
+}
+
+// Global describes global environment specs
+type Global struct {
+	Spec *Spec `yaml:"spec"`
+}
+
+// Spec contains the spec of a given environment/instance
+type Spec struct {
+	Kubernetes      Kubernetes        `yaml:"kubernetes"`
+	Secrets         []*v2e.SecretItem `yaml:"secrets"`
+	EnvironmentVars []*EnvironmentVar `yaml:"env"`
+}
+
+// Kubernetes describes the Kubernetes configuration to use
+type Kubernetes struct {
+	ServiceAccount string `yaml:"serviceAccount"`
+	Cluster        string `yaml:"cluster"`
+}
+
+// Environment describes a deployment environment (i.e. dev, stage, prod, etc.)
+type Environment struct {
+	Name        string      `yaml:"name"`
+	Spec        *Spec       `yaml:"spec"`
+	Instances   []*Instance `yaml:"instances"`
+	instanceMap map[string]int
+}
+
+// Instance describes an instance of a deployment within an environment (i.e. us-west-2 for env prod)
+type Instance struct {
+	Name string `yaml:"name"`
+	Spec *Spec  `yaml:"spec"`
+}
+
+// EnvironmentVar describes a shell env var to be injected into the deployment environment
+type EnvironmentVar struct {
+	Name  string `yaml:"name"`
+	Value string `yaml:"value"`
+}
+
+// parseConfig opens the deployment config file and ensures it is valid
+func (d *Deploy) parseConfig() {
+
+	d.config = Config{}
+
+	configFile := d.stim.GetConfig("deploy.file")
+
+	if configFile == "" {
+		setConfigDefault(&configFile, defaultConfigFile)
+		d.log.Debug("Deployment file not specified, using {}", defaultConfigFile)
+	}
+
+	_, err := os.Stat(configFile)
+	if err != nil && !os.IsExist(err) {
+		d.log.Fatal("No deployment config file exists at: {}", configFile)
+	}
+
+	contentstring, err := ioutil.ReadFile(configFile)
+	if err != nil {
+		d.log.Fatal("Deployment config file could not be read: {}", err)
+	}
+
+	if ok, err := utils.IsYaml(contentstring); !ok {
+		d.log.Fatal("Deployment config file ({}) is not valid YAML: {}", configFile, err)
+	}
+
+	err = yaml.Unmarshal([]byte(contentstring), &d.config)
+	if err != nil {
+		d.log.Fatal("Error parsing deployment config {}", err)
+	}
+
+	d.config.configFilePath = configFile
+
+	d.processConfig()
+
+}
+
+// processConfig ensures that the deployment config is valid
+func (d *Deploy) processConfig() {
+
+	// Set defaults
+	setConfigDefault(&d.config.Deployment.Container.Repo, defaultContainerRepo)
+	setConfigDefault(&d.config.Deployment.Container.Tag, defaultContainerTag)
+	setConfigDefault(&d.config.Deployment.Directory, defaultDeployDirectory)
+	setConfigDefault(&d.config.Deployment.Script, defaultDeployScript)
```

Comment:
I am interested in an alternative ways of doing this as we kinda have this issue in other packages.  I tried creating a default instance of the struct and merging it with the one created from the config file but merging complex data structures might not be the way to go.  

In the scenario above, those variables are not exactly locally scoped (since they're part of the main config object) and I figured passing around a pointer is a bit more efficient than passing the data by value; copying the data in and out of a function over and over.

I suppose we could have a single function that takes care of all of it since this is all taking place within the same struct.  Like so:
```
d.setDefault()

func (d *Deploy) setDefaults() {
	if d.config.Deployment.Directory == "" {
		d.config.Deployment.Directory = defaultDeployDirectory
	}
	if d.config.Deployment.Script == "" {
		d.config.Deployment.Script = defaultDeployScript
	}
	if d.config.Deployment.Container.Repo == "" {
		d.config.Deployment.Container.Repo = defaultContainerRepo
	}
	if d.config.Deployment.Container.Tag == "" {
		d.config.Deployment.Container.Tag = defaultContainerTag
	}
}
```

What do you think?

Reply:
I do like the idea of a function that just does that much better.  That was actually my first thought when I saw what it was doing.  It would be nice if the yaml Marshaling stuff could also set defaults.

When I say out of scope its more that the d.config.* pointer is passed into a function that otherwise could never have access to it since its not part of the Deploy struct.  Which is more efficient is hard to say it really depends on how golang compiles down and if/how it deals with in-lining functions, either way I would expect either method is only a few calls away from each other.

To me its more about the lack of visibility for a simple operation.  Your going into another context and modifying/assigning things in the current context w/o easy visibility into it, granted there is reason to do that, but in general I think anything modifying/assigning d.config.* should only be done in the Deploy Struct.